src/posix/platform/netif.cpp

```diff
@@ -0,0 +1,519 @@
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the platform network on Linux.
+ */
+
+#include "platform-posix.h"
+
+#include <arpa/inet.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <ifaddrs.h>
+#if __linux__
+#include <linux/if_tun.h>
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+#endif
+#include <net/if.h>
+#include <net/if_arp.h>
+#include <stdio.h>
+#include <sys/ioctl.h>
+#include <sys/select.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <openthread/instance.h>
+#include <openthread/ip6.h>
+#include <openthread/message.h>
+
+#include "common/code_utils.hpp"
+#include "common/logging.hpp"
+#include "net/ip6_address.hpp"
+
+#if OPENTHREAD_ENABLE_PLATFORM_NETIF
+
+#ifndef OPENTHREAD_POSIX_TUN_DEVICE
+#define OPENTHREAD_POSIX_TUN_DEVICE "/dev/net/tun"
+#endif // OPENTHREAD_TUN_DEVICE
+
+// from linux/ipv6.h
+struct in6_ifreq
+{
+    struct in6_addr ifr6_addr;
+    __u32           ifr6_prefixlen;
+    int             ifr6_ifindex;
+};
+
+static otInstance * sInstance  = NULL;
+static int          sTunFd     = -1; ///< Used to exchange IPv6 packets.
+static int          sIpFd      = -1; ///< Used to manage IPv6 stack on Thread interface.
+static int          sNetlinkFd = -1; ///< Used to receive netlink events.
+static unsigned int sTunIndex  = 0;
+static char         sTunName[IFNAMSIZ];
+
+static const size_t  kMaxIp6Size            = 1536;
+static const uint8_t kMulticastPrefixLength = 128;
+
+static void UpdateUnicast(otInstance *aInstance, const otIp6Address &aAddress, uint8_t aPrefixLength, bool aIsAdded)
+{
+    struct in6_ifreq ifr6;
+    otError          error = OT_ERROR_NONE;
+
+    assert(sInstance == aInstance);
+
+    VerifyOrExit(sIpFd > 0, error = OT_ERROR_INVALID_STATE);
+
+    memcpy(&ifr6.ifr6_addr, &aAddress, sizeof(ifr6.ifr6_addr));
+
+    ifr6.ifr6_ifindex   = sTunIndex;
+    ifr6.ifr6_prefixlen = aPrefixLength;
+
+    VerifyOrExit(ioctl(sIpFd, (aIsAdded ? SIOCSIFADDR : SIOCDIFADDR), &ifr6) == 0, perror("ioctl");
+                 error = OT_ERROR_FAILED);
```

Comment:
nit:
how about adding a macro to let reader read the code more easier:
```
#define OT_PERROR(ERR_STRING, ERR_CODE) perror(ERR_STRING); error= ERR_CODE;
VerifyOrExit(ioctl(sIpFd, (aIsAdded ? SIOCSIFADDR : SIOCDIFADDR), &ifr6) == 0, OT_PERROR("ioctl", OT_ERROR_FAILED));
```

Reply:
When I try applying this macro, I found it not self-contained: it requires there's a local `error` variable, so I tend to keep it as is. Thoughts?